### DISCUSSION TITLE
Reducing the size (and complexity) of the template source

### DISCUSSION THREAD

**Discussion by joaomlourenco** [joaomlourenco/novathesis#302] [Ideas]
The template contains multiple files and folders.  One of these folders, `NOVAthesisFiles/Schools` contains the configuration files for the many schools supported by the template, and amounts to 21 MB.

But each individual is from a single School and does not need all those configuration files for the other Schools.  I foresee two ways to deal with this:

1. Provide just one or two Schools by default, and allow people to download and install additional Schools (their own School).
  - (+) Smaller download
  - (+) Less space on disk
  - (–) Less proficient users may find it hard to download/install the additional School's configuration files.
2. Provide a script that deletes all thee configuration files, except from one (selected) School.
  - (+) Deleting the extra files becomes optional
  - (+) Much easier to deal with
  - (–) Download is still large and initial disk space as well

What are your feelings concerning the options (1) and (2)?

**Comment by dcaneco** [comment]
Personally the size of the download is not an issue. I would like to have a script to delete because I've deleted some by hand but I know I missed a few on the way.
Thanks a lot

**Reply by joaomlourenco** [reply]
@dcaneco, thank you for sharing your opinion.   I'll work on the windows and linux/macOS scripts soon.

**Comment by joaomlourenco** [comment]
Interesting… no one else cares about this!  I wonder why…
Still in my to do list!

**Comment by GilOliveira** [comment]
My 2 cents is that this issue is, at least to me, not a priority. It does seem like a nice feature to save some space, but I personally never had any issues with space on my computer and on free-tier git services to sync all those files. I'm not particularly tight in disk space so it can just be me.

I think there's, however, some merit to the effort. I'm not very knowledgeable in the inner-workings of the template, but if this effort leads to the template being more modular, it seems to me that it would help folks who are interested in developing for their own schools to only need to worry about a particular set of sub-files and just drop them in the template almost like a plugin.

**Reply by joaomlourenco** [reply]
@GilOliveira, thanks for your input.  👍

Just to followup on your suggestion of modularity…  The template already works that way, it is modular and adding a new School is like adding a plugin (usually all the necessary customization is localized in a single folder).   But that is not explicit…  and I am afraid that, every time I add support for yet another school to the NOVAthesis distribution, it becomes larges and larger.   I am afraid one of these days Overleaf will say "your template is too large to be hosted in Overleaf". 😟

**Reply by GilOliveira** [reply]
Thanks, professor! That was my expectation! I'm fortunate enough to be doing research at NOVA so first-party support was never much of an issue so I never actually looked at the source files.

**Comment by tpdsantos** [comment]
Why not create a script that can download and delete those school files that are unnecessary? 
I would go for the 2) premise, but instead of deleting the files, interactively ask which school(s) does the user want to use, and only download the corresponding files. This also helps non-experienced users to re-download accidentally deleted files or easily switch school templates, since all the schools that the user did not specify can be easily removed from the local directory.

I think this approach is particularly good because we can skip downloading the `.git` folder, which is always the heaviest folder in any GitHub repository.

I would propose some sort of an INSTALL script that downloads the core template files and interactively asks the user:

- which school(s) do you want to use? (show list of supported schools)
- your `Schools` folder is not empty, what do you want to do? (replace, add, overwrite)

And probably other things that I'm forgetting.

**Reply by joaomlourenco** [reply]
@tpdsantos, The problem with this approach is that there is Windows.  One needs to write and maintain two versions of the script, one for Linux and macOS and another for WIndows.  I'm not sure I want to go on that path for complex scripts.  

Is there any solution that would work out of the box in both Linux/macOS and Windows?  Maybe something in Perl or Python, but it still would require the user to instal one of those systems.

What do you think?

**Reply by tpdsantos** [reply]
@joaomlourenco you have a point, it would have to be both scripts for sure... I don't think Python would be an option, since the easiest installation for beginners would be anaconda, which is around 300MB. About Perl I have no idea since I never used it, is it a small installation?

**Reply by joaomlourenco** [reply]
@tpdsantos, I believe that the TeX-Live (and probably MikTeX) LaTeX installation requires perl for some of the scripts (e.g., latexmk, makeglossaries, etc), so perl feels as a better candidate.  But… still… not sure.  I have to create a VM with windows and install TeXLive and MikTeX and see if perl is available.  

But even if perl is available, I'm afraid we will need libraries (e.g., to download files) that may not be available.   We need the help of an expert in perl scripting…  :D

**Reply by tpdsantos** [reply]
Yeah, you're probably right, Windows scripting is always a pain 😅

**Reply by GilOliveira** [reply]
I just downloaded the entire repository through DQ FCT NOVA's wired connection... and considering the time it took... I second the opinion of not having to download everything 😅

Not very familiar with [GitHub Actions](https://github.com/features/actions) but perhaps there's a way of configuring it so it splits every new release into modules (like a core one plus a module for every school) and people just download what they need from the [releases section](https://github.com/joaomlourenco/novathesis/releases), maybe?

**Reply by tpdsantos** [reply]
To be fair, the entire module is not that big. I think a good first approach would be to just get a script that downloads everything but the `.git` folder, which takes more than 70% of the entire project. With that, the improvement would be immense!

**Comment by joaomlourenco** [comment]
@tpdsantos, downloading everything except the .git folder is what you get when you download the zip file.  😃
Also, if you do
```shell
git clone --depth=1 https://github.com/joaomlourenco/novathesis.git
```
the download will be much smaller (≈70 MB vs ≈400 MB).

**Reply by tpdsantos** [reply]
really?? that I didn't know... I probably cloned it using git, I can't remember xD
